uaclient/files.py

```diff
@@ -0,0 +1,344 @@
+        root_directory: str = defaults.DEFAULT_DATA_DIR,
+        private: bool = True,
+    ):
+        self._root_directory = root_directory
+        self._file_name = name
+        self._is_private = private
+        self._path = None
+
+    @property
+    def path(self) -> str:
+        if not self._path:
+            self._path = os.path.join(
+                self._root_directory, self._file_name
+            )  # type: ignore
+        return self._path  # type: ignore
+
+    @property
+    def is_private(self) -> bool:
+        return self._is_private
+
+    def write(self, content):
+        file_mode = (
+            defaults.ROOT_READABLE_MODE
+            if self.is_private
+            else defaults.WORLD_READABLE_MODE
+        )
+        self.create_dir()
+        if not isinstance(content, str):
+            content = json.dumps(content, cls=util.DatetimeAwareJSONEncoder)
+        util.write_file(self.path, content, file_mode)
+
+    def read(self) -> Optional[str]:
+        content = None
+        try:
+            content = util.load_file(self.path)
+        except FileNotFoundError:
+            if not os.path.exists(self.path):
+                LOG.debug("File does not exist: {}".format(self.path))
+        return content
+
+    def delete(self):
+        util.remove_file(self.path)
+
+    def create_dir(self):
+        file_dir = os.path.dirname(self.path)
+        if not os.path.exists(file_dir):
+            os.makedirs(file_dir)
+            if self.is_private:
+                os.chmod(file_dir, 0o700)
+
+
+class MachineTokenFile:
+    def __init__(
+        self,
+        root_directory: str = defaults.DEFAULT_DATA_DIR,
+        root_mode: bool = True,
+        machine_token_overlay_path: Optional[str] = None,
+    ):
+        file_name = defaults.MACHINE_TOKEN_FILE
+        self.is_root = root_mode
+        self.private_file = UAFile(
+            file_name, root_directory + defaults.PRIVATE_SUBDIR
+        )
+        self.public_file = UAFile(file_name, root_directory)
+        self.machine_token_overlay_path = machine_token_overlay_path
+        self._machine_token = None
+        self._entitlements = None
+        self._contract_expiry_datetime = None
+
+    def write(self, content):
+        """Update the machine_token file for both pub/private files"""
+        if self.is_root:
+            self.private_file.write(content)
+            content = PublicMachineTokenData.from_dict(content).to_dict()
+            self.public_file.write(content)
+
+            self._machine_token = None
+            self._entitlements = None
+        else:
+            raise exceptions.NonRootUserError()
+
+    def delete(self):
+        """Delete both pub and private files"""
+        if self.is_root:
+            self.public_file.delete()
+            self.private_file.delete()
+
+            self._machine_token = None
+            self._entitlements = None
+            self.public_file = None
+            self.private_file = None
```

Comment:
We may want to do something like
```
machine_token_file.delete()
machine_token_file.write(new_content)
```

so I don't think we can set these fields to None

Reply:
Makes sense, did not see that. Since we only instantiate them in constructor.